### PR TITLE
Issue #7: Settings and expiry

### PR DESCRIPTION
**Pull request by jaypha**
Add settings page
Add expiry task

### CODE REVIEW

File: collector/database/classes/task/expire_metrics.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace cltr_database\task;
+
+use cltr_database\lib;
+
+class expire_metrics extends \core\task\scheduled_task {
```
Comment (brendanheywood) [inline collector/database/classes/task/expire_metrics.php:30]: purely to be more 'moodley' can you rename this to xyz_cleanup_task ie metrics_cleanup_task, de facto convention for this:

https://github.com/moodle/moodle/tree/master/lib/classes/task

File: collector/database/classes/task/expire_metrics.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    public function get_name() {
+        return get_string('task_expire_metrics', 'cltr_database');
+    }
+
+    public function execute() {
+        global $DB;
+
+        $cutoff = time() - lib::get_metric_expiry();
```
Comment (brendanheywood) [inline collector/database/classes/task/expire_metrics.php:39]: bonus points might be nice to round this to the nearest day so we don't end up with half a days worth of data left over

File: collector/database/settings.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+if ($hassiteconfig) {
+    if ($ADMIN->fulltree) {
+        $settings->add(
+            new admin_setting_configtext(
```
Comment (brendanheywood) [inline collector/database/settings.php:31]: Would be a little nicer as a admin_setting_configduration

File: settings.php [hunk 4]
```diff
@@ -56,4 +56,8 @@
             new tool_cloudmetrics\admin_setting_manage_metrics()
         );
     }
+
+    foreach (cltr::get_enabled_plugins() as $plugin) {
```
Comment (brendanheywood) [inline settings.php:60]: It is often useful to be able to configure plugins before they are enabled like auth plugins do

File: collector/database/classes/task/expire_metrics.php (hunk 1 above)
Comment (jaypha) [reply collector/database/classes/task/expire_metrics.php]: Done

File: settings.php (hunk 4 above)
Comment (jaypha) [reply settings.php]: Done

File: collector/database/settings.php (hunk 3 above)
Comment (jaypha) [reply collector/database/settings.php]: Done

File: collector/database/classes/task/expire_metrics.php (hunk 2 above)
Comment (jaypha) [reply collector/database/classes/task/expire_metrics.php]: Putting this into the 'too hard' basket for now.
Besides, you still haven't told me what I can spend my bonus points on. 🙂

File: collector/database/classes/lib.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * @package   cltr_database
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class lib {
+
+    // TODO: Is there a better way to get this?
+    /** @var string Root directory of the plugin. */
+    const PLUGIN_DIR = '/admin/tool/cloudmetrics/collector/database/';
```
Comment (brendanheywood) [inline collector/database/classes/lib.php:31]: you are probably after something like this:

```
lib/classes/plugininfo/local.php:66:        if (file_exists($this->full_path('settings.php'))) {
lib/classes/plugininfo/local.php:67:            include($this->full_path('settings.php'));
lib/classes/plugininfo/local.php-68-        }
```

File: collector/database/classes/lib.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    /** @var string Name of database table. */
+    const TABLE = 'cltr_database_metrics';
+
+    /** @var int Number of seconds in one day. */
+    const SECS_IN_DAY = 86400;
```
Comment (brendanheywood) [inline collector/database/classes/lib.php:37]: there is a core define:

lib/moodlelib.php:49:define('DAYSECS', 86400);

File: collector/database/classes/task/expire_metrics.php (hunk 2 above)
Comment (brendanheywood) [reply collector/database/classes/task/expire_metrics.php]: split into https://github.com/catalyst/moodle-tool_cloudmetrics/issues/23

File: collector/database/classes/lib.php (hunk 6 above)
Comment (jaypha) [reply collector/database/classes/lib.php:37]: Thanks. Must have been on my blind spot when looking for it.

**Pull request merged**